strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/layout/pypi/PypiArtifactControllerTest.java

```diff
@@ -0,0 +1,272 @@
+    public void testUploadPackageFlow(@PypiTestRepository(repositoryId = REPOSITORY_RELEASES, storageId = REPOSITORY_STORAGE) Repository repository,
+                                      @PypiTestArtifact(repositoryId = REPOSITORY_RELEASES, storageId = REPOSITORY_STORAGE, id = "hello_world_pypi", versions = "1.3") Path packagePath)
+    {
+        final String storageId = repository.getStorage().getId();
+        final String repositoryId = repository.getId();
+
+        final RepositoryPath repositoryPath = (RepositoryPath) packagePath.normalize();
+
+        String url = getContextBaseUrl() + "/storages/{storageId}/{repositoryId}";
+
+        // Upload with Invalid action
+        mockMvc.contentType(MediaType.MULTIPART_FORM_DATA_VALUE)
+               .multiPart("filetype", "sdist")
+               .multiPart(":action", "File_Upload")
+               .multiPart("name", "hello-world-pipy")
+               .multiPart("metadata_version", "1.0")
+               .multiPart("content", repositoryPath.toFile())
+               .when()
+               .post(url, storageId, repositoryId)
+               .then()
+               .log()
+               .all()
+               .statusCode(HttpStatus.BAD_REQUEST.value())
+               .contentType(ContentType.TEXT);
+
+        // Upload with Invalid filetype
+        mockMvc.contentType(MediaType.MULTIPART_FORM_DATA_VALUE)
+               .multiPart("filetype", "wheel")
+               .multiPart(":action", "file_upload")
+               .multiPart("name", "hello-world-pipy")
+               .multiPart("metadata_version", "1.0")
+               .multiPart("content", repositoryPath.toFile())
+               .when()
+               .post(url, storageId, repositoryId)
+               .then()
+               .log()
+               .all()
+               .statusCode(HttpStatus.BAD_REQUEST.value())
+               .contentType(ContentType.TEXT);
+
+        // Valid pypi package upload
+        mockMvc.contentType(MediaType.MULTIPART_FORM_DATA_VALUE)
+               .multiPart("filetype", "sdist")
+               .multiPart(":action", "file_upload")
+               .multiPart("name", "hello-world-pipy")
+               .multiPart("metadata_version", "1.0")
+               .multiPart("content", repositoryPath.toFile())
+               .when()
+               .post(url, storageId, repositoryId)
+               .then()
+               .log()
+               .all()
+               .statusCode(HttpStatus.OK.value())
+               .contentType(ContentType.TEXT)
+               .body(Matchers.containsString("The artifact was deployed successfully."));
+    }
+
+    @Test
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class,
+                  ArtifactManagementTestExecutionListener.class })
+    public void testDownloadPackageRedirectionFlow(@PypiTestRepository(repositoryId = REPOSITORY_RELEASES, storageId = REPOSITORY_STORAGE) Repository repository)
```

Comment:
@anki2189 please use `@PypiTestArtifact` here to prepare test artifact that you want to download.

Reply:
Sure. Noted